engine/ext_make3dgrains.go

```diff
@@ -0,0 +1,158 @@
+package engine
+
+import (
+	"math"
+	"math/rand"
+)
+
+func init() {
+	DeclFunc("ext_make3dgrains", Voronoi3d, "3D Voronoi tesselation over shape (grain size, starting region number, num regions, shape, seed)")
+}
+
+func Voronoi3d(grainsize float64, startRegion int, numRegions int, inputShape Shape, seed int) {
+	Refer("Lel2014")
+	SetBusy(true)
+	defer SetBusy(false)
+
+	t := newTesselation3d(grainsize, numRegions, int64(seed), startRegion, inputShape)
+	regions.hist = append(regions.hist, t.RegionOf)
+	regions.render(t.RegionOf)
+}
+
+type tesselation3d struct {
+	grainsize   float64
+	maxRegion   int
+	seed        int64
+	rnd         *rand.Rand
+	startRegion int
+	shape       Shape			//Shape of the tesselated region
+	centers     []center3d			//List of Voronoi centers
+}
+
+
+// Stores location of each Voronoi center
+type center3d struct {
+	x, y, z   float64 // center position (m)
+	region byte    // region for all cells near center
+}
+
+// Stores location of each cell
+type cellLocs struct {x, y, z float64}
+
+// nRegion exclusive
+func newTesselation3d(grainsize float64, nRegion int, seed int64, startRegion int, inputShape Shape) *tesselation3d {
+	t := tesselation3d{grainsize,
+		nRegion,
+		//make(map[int3d][]center3d),
+		seed,
+		rand.New(rand.NewSource(0)),
```

Comment:
shouldn't seed be passed to NewSource?

Reply:
Yes, good catch. I have removed the seed from the tesselation3d type, and just passed the seed directly to NewSource.